examples/headless/src/main.rs

```diff
@@ -0,0 +1,248 @@
+use wgpu::{
+    util::initialize_adapter_from_env_or_default, Backends, BufferDescriptor, BufferUsages,
+    CommandEncoderDescriptor, DeviceDescriptor, Extent3d, ImageCopyBuffer, Instance, Limits,
+    TextureDescriptor, TextureFormat, TextureUsages,
+};
+
+fn main() -> Result<()> {
+    #[cfg(not(target_arch = "wasm32"))]
+    env_logger::init();
+    let args = Args::parse();
+    let scenes = args.args.select_scene_set(|| Args::command())?;
+    if let Some(scenes) = scenes {
+        let mut scene_idx = None;
+        for (idx, scene) in scenes.scenes.iter().enumerate() {
+            if scene.config.name.eq_ignore_ascii_case(&args.scene) {
+                if let Some(scene_idx) = scene_idx {
+                    eprintln!("Scene names conflict, skipping scene {idx} (instead rendering {scene_idx})");
+                } else {
+                    scene_idx = Some(idx);
+                }
+            }
+        }
+        let scene_idx = match scene_idx {
+            Some(idx) => idx,
+            None => {
+                let parsed = args.scene.parse::<usize>().context(format!(
+                    "'{}' didn't match any scene, trying to parse as index",
+                    args.scene
+                ))?;
+
+                if !(parsed < scenes.scenes.len()) {
+                    if scenes.scenes.len() == 0 {
+                        bail!("Cannot select a scene, as there are no scenes")
+                    }
+                    bail!(
+                        "{parsed} doesn't fit in scenes (len {})",
+                        scenes.scenes.len()
+                    );
+                }
+                parsed
+            }
+        };
+        if args.print_scenes {
+            println!("Available scenes:");
+
+            for (idx, scene) in scenes.scenes.iter().enumerate() {
+                println!(
+                    "{idx}: {}{}{}",
+                    scene.config.name,
+                    if scene.config.animated {
+                        " (animated)"
+                    } else {
+                        ""
+                    },
+                    if scene_idx == idx { " (selected)" } else { "" }
+                );
+            }
+            return Ok(());
+        }
+        pollster::block_on(render(scenes, scene_idx, &args))?;
+    }
+    Ok(())
+}
+
+async fn render(mut scenes: SceneSet, index: usize, args: &Args) -> Result<()> {
+    let instance = Instance::new(wgpu::InstanceDescriptor {
+        backends: wgpu::Backends::PRIMARY,
+        dx12_shader_compiler: wgpu::Dx12Compiler::Fxc,
+    });
+    let adapter = initialize_adapter_from_env_or_default(&instance, Backends::PRIMARY, None)
+        .await
+        .ok_or_else(|| anyhow!("Failed to intiialise adapter"))?;
+    let (device, queue) = adapter
+        .request_device(
+            &DeviceDescriptor {
+                label: Some("Vello Headless"),
+                features: wgpu::Features::TIMESTAMP_QUERY | wgpu::Features::CLEAR_TEXTURE,
```

Comment:
> features: wgpu::Features::TIMESTAMP_QUERY | wgpu::Features::CLEAR_TEXTURE,

I think this should get filtered through the current adapter's available features (e.g. `adapter.features() & (wgpu::Features::TIMESTAMP_QUERY | wgpu::Features::CLEAR_TEXTURE)`). I believe TIMESTAMP_QUERY is not available on my M1 Pro and the example terminates with `Error in Adapter::request_device: unsupported features were requested: TIMESTAMP_QUERY` unless I make this change. I noticed that this is what `vello::util::RenderContext::new_device` does internally.

What do you think about creating the device and queue via `vello::util::RenderContext` here? You could add a public method to `RenderContext` that initializes and returns a device handle without requiring a surface. That would address this issue while reducing code duplication and make this code consistent with the `with_winit` example.